Index: sauerbraten_unix
===================================================================
--- sauerbraten_unix	(revision 6291)
+++ sauerbraten_unix	(working copy)
@@ -1,4 +1,10 @@
 #!/bin/sh
+
+if [ ${XDG_SESSION_TYPE} == wayland ]
+then
+  export SDL_VIDEODRIVER=wayland
+fi
+
 # SAUER_DATA should refer to the directory in which Sauerbraten data files are placed.
 #SAUER_DATA=~/sauerbraten
 #SAUER_DATA=/usr/local/sauerbraten
