Index: sauerbraten_unix
===================================================================
--- sauerbraten_unix	(revision 6291)
+++ sauerbraten_unix	(working copy)
@@ -10,12 +10,18 @@
 #SAUER_DATA=/usr/local/sauerbraten
 SAUER_DATA=.
 
+# SAUER_USER_DIR should refer to the directory in which user files are placed.
+SAUER_USER_DIR="${HOME}/.sauerbraten_svn"
+
 # SAUER_BIN should refer to the directory in which Sauerbraten executable files are placed.
 SAUER_BIN=${SAUER_DATA}/bin_unix
 
+# SAUER_USER_BIN should refer to the directory in which custom executable files are placed.
+SAUER_USER_BIN=${SAUER_USER_DIR}/bin_unix
+
 # SAUER_OPTIONS contains any command line options you would like to start Sauerbraten with.
 #SAUER_OPTIONS="-f"
-SAUER_OPTIONS="-q${HOME}/.sauerbraten"
+SAUER_OPTIONS="-q${SAUER_USER_DIR}"
 
 # SYSTEM_NAME should be set to the name of your operating system.
 #SYSTEM_NAME=Linux
@@ -50,6 +56,12 @@
   ;;
 esac
 
+if [ -x ${SAUER_USER_BIN}/native_client ]
+then
+  SYSTEM_NAME=native_
+  MACHINE_NAME=
+fi
+
 if [ -x ${SAUER_BIN}/native_client ]
 then
   SYSTEM_NAME=native_
@@ -56,9 +68,13 @@
   MACHINE_NAME=
 fi
 
-if [ -x ${SAUER_BIN}/${SYSTEM_NAME}${MACHINE_NAME}client ]
+if [ -x ${SAUER_USER_BIN}/${SYSTEM_NAME}${MACHINE_NAME}client ]
 then
   cd ${SAUER_DATA}
+  exec ${SAUER_USER_BIN}/${SYSTEM_NAME}${MACHINE_NAME}client ${SAUER_OPTIONS} "$@"
+elif [ -x ${SAUER_BIN}/${SYSTEM_NAME}${MACHINE_NAME}client ]
+then
+  cd ${SAUER_DATA}
   exec ${SAUER_BIN}/${SYSTEM_NAME}${MACHINE_NAME}client ${SAUER_OPTIONS} "$@"
 else
   echo "Your platform does not have a pre-compiled Sauerbraten client."
