diff --git src/engine/server.cpp src/engine/server.cpp
index 745d655..fab466d 100644
--- src/engine/server.cpp
+++ src/engine/server.cpp
@@ -203,6 +203,7 @@ int setclientrealip(int n, uint ip)
     return enet_address_get_host_ip(&clients[n]->real, clients[n]->hostname, strlen("xxx.xxx.xxx.xxx"));
 }
 const char *getclienthostname(int n) { return clients.inrange(n) && clients[n]->type==ST_TCPIP ? clients[n]->hostname : NULL; }
+ICOMMAND(getip, "i", (int *cn), stringret(newstring(getclienthostname(*cn))));

 void sendpacket(int n, int chan, ENetPacket *packet, int exclude)
 {
diff --git src/fpsgame/server.cpp src/fpsgame/server.cpp
index 4faba44..c05121e 100644
--- src/fpsgame/server.cpp
+++ src/fpsgame/server.cpp
@@ -110,6 +110,34 @@ namespace server
         bannedips.add(b);
     }

+    ICOMMAND(ban, "si", (char *name, int *minutes),
+    {
+        uint ip;
+        if(inet_pton(AF_INET, name, &ip) < 0) return;
+        addban(ip, *minutes * 60*1000);
+    });
+    ICOMMAND(unban, "s", (char *name),
+    {
+        uint ip;
+        if(inet_pton(AF_INET, name, &ip) < 0) return;
+        loopv(bannedips) if(bannedips[i].ip==ip) { bannedips.remove(i); }
+    });
+    ICOMMAND(listbans, "", (),
+    {
+        if(!bannedips.length()) { stringret(newstring("no bans")); return; }
+        char *msg = newstring(1024);
+        loopv(bannedips)
+        {
+            string banline = "\n";
+            if(inet_ntop(AF_INET, &bannedips[i].ip, banline+1, strlen("xxx.xxx.xxx.xxx")+1) == NULL) continue;
+            conoutf("%s", banline);
+            int expsecs = (bannedips[i].expire - totalmillis) / 1000;
+            concformatstring(banline, " (%dh %dm %ds)", expsecs/(60*60), expsecs/60, expsecs%60);
+            concatstring(msg, banline, 1024);
+        };
+        stringret(msg);
+    });
+
     vector<clientinfo *> connects, clients, bots;

     void kickclients(uint ip, clientinfo *actor = NULL, int priv = PRIV_NONE)
@@ -1241,6 +1270,7 @@ namespace server
                 if(reason && reason[0]) sendservmsgf("%s kicked %s because: %s", kicker, colorname(vinfo), reason);
                 else sendservmsgf("%s kicked %s", kicker, colorname(vinfo));
                 uint ip = getclientip(victim);
+                logoutf("getclientip returned %u for %s (cn %d)", ip, colorname(vinfo), victim);
                 addban(ip, 4*60*60000);
                 kickclients(ip, ci, priv);
             }
