Index: src/engine/command.cpp
===================================================================
--- src/engine/command.cpp	(revision 6490)
+++ src/engine/command.cpp	(working copy)
@@ -2365,7 +2365,7 @@
 {
     string s;
     copystring(s, cfgfile);
-    char *buf = loadfile(path(s), NULL);
+    char *buf = loadfile(path(s), NULL, true, strlen(s)>3 && !strncmp(s+strlen(s)-3, ".gz", 3));
     if(!buf)
     {
         if(msg) conoutf(CON_ERROR, "could not read \"%s\"", cfgfile);
Index: src/shared/stream.cpp
===================================================================
--- src/shared/stream.cpp	(revision 6490)
+++ src/shared/stream.cpp	(working copy)
@@ -1237,9 +1237,9 @@
     return utf8;
 }

-char *loadfile(const char *fn, size_t *size, bool utf8)
+char *loadfile(const char *fn, size_t *size, bool utf8, bool gz)
 {
-    stream *f = openfile(fn, "rb");
+    stream *f = gz ? opengzfile(fn, "rb") : openfile(fn, "rb");
     if(!f) return NULL;
     stream::offset fsize = f->size();
     if(fsize <= 0) { delete f; return NULL; }
Index: src/shared/tools.h
===================================================================
--- src/shared/tools.h	(revision 6490)
+++ src/shared/tools.h	(working copy)
@@ -1368,7 +1368,7 @@
 extern stream *opentempfile(const char *filename, const char *mode);
 extern stream *opengzfile(const char *filename, const char *mode, stream *file = NULL, int level = Z_BEST_COMPRESSION);
 extern stream *openutf8file(const char *filename, const char *mode, stream *file = NULL);
-extern char *loadfile(const char *fn, size_t *size, bool utf8 = true);
+extern char *loadfile(const char *fn, size_t *size, bool utf8 = true, bool gz = false);
 extern bool listdir(const char *dir, bool rel, const char *ext, vector<char *> &files);
 extern int listfiles(const char *dir, const char *ext, vector<char *> &files);
 extern int listzipfiles(const char *dir, const char *ext, vector<char *> &files);
