diff --git src/fpsgame/client.cpp src/fpsgame/client.cpp
index 8700785..48dd292 100644
--- src/fpsgame/client.cpp
+++ src/fpsgame/client.cpp
@@ -957,6 +957,8 @@ namespace game
     void connectattempt(const char *name, const char *password, const ENetAddress &address)
     {
         copystring(connectpass, password);
+        extern char *filterservers;
+        setsvar("filterservers", "");
     }

     void connectfail()
diff --git src/fpsgame/fps.cpp src/fpsgame/fps.cpp
index 141739f..ce35f82 100644
--- src/fpsgame/fps.cpp
+++ src/fpsgame/fps.cpp
@@ -1276,8 +1276,15 @@ namespace game
         return (n>=MM_START && size_t(n-MM_START)<sizeof(mastermodeicons)/sizeof(mastermodeicons[0])) ? mastermodeicons[n-MM_START] : unknown;
     }

+    SVAR(filterservers, "");
+
     bool serverinfoentry(g3d_gui *g, int i, const char *name, int port, const char *sdesc, const char *map, int ping, const vector<int> &attr, int np)
     {
+        if(*filterservers)
+            if(!cubecasefind(sdesc, filterservers) &&
+               !cubecasefind(map, filterservers) &&
+               (attr.length()<2 || !cubecasefind(server::modename(attr[1], ""), filterservers))) return false;
+
         #define leftjustified(elem)   justified(elem,true,false)
         #define rightjustified(elem)  justified(elem,true,true)

