Index: src/fpsgame/ctf.h
===================================================================
--- src/fpsgame/ctf.h	(revision 6502)
+++ src/fpsgame/ctf.h	(working copy)
@@ -1267,10 +1267,25 @@
 		}
 		return false;
 	}
+
+    fpsent *getflagholder(const char *teamname) {
+        if(m_hold) return flags.length() == 1 ? flags[0].owner : NULL;
+        if(!strlen(teamname)) return NULL;
+        loopv(flags)
+        {
+            flag &f = flags[i];
+            if(f.team == ctfteamflag(teamname)) return f.owner;
+        }
+        return NULL;
+    }
 };

 extern ctfclientmode ctfmode;
 ICOMMAND(dropflag, "", (), { ctfmode.trydropflag(); });
+ICOMMAND(getflagholdercn, "s", (const char *teamname), {
+    fpsent *h = ctfmode.getflagholder(teamname);
+    intret(h ? h->clientnum : -1);
+});

 #endif

