Index: src/fpsgame/ctf.h
===================================================================
--- src/fpsgame/ctf.h	(revision 6502)
+++ src/fpsgame/ctf.h	(working copy)
@@ -1267,10 +1267,17 @@
 		}
 		return false;
 	}
+
+    bool hasflag(fpsent *d) {
+        if(!d) return false;
+        loopv(flags) if(flags[i].owner == d) return true;
+        return false;
+    }
 };

 extern ctfclientmode ctfmode;
 ICOMMAND(dropflag, "", (), { ctfmode.trydropflag(); });
+ICOMMAND(hasflag,  "i", (int *cn), intret(ctfmode.hasflag(getclient(*cn))));

 #endif

