Index: src/fpsgame/scoreboard.cpp
===================================================================
--- src/fpsgame/scoreboard.cpp	(revision 6479)
+++ src/fpsgame/scoreboard.cpp	(working copy)
@@ -483,7 +483,7 @@
     void drawhudscore(int w, int h)
     {
         int numgroups = groupplayers();
-        if(!numgroups) return;
+        if(!numgroups || (m_teammode ? numgroups : groups[0]->players.length()) < 2) return;

         scoregroup *g = groups[0];
         int score = INT_MIN, score2 = INT_MIN;
