diff --git src/shared/zip.cpp src/shared/zip.cpp
index c35fa87..341e261 100644
--- src/shared/zip.cpp
+++ src/shared/zip.cpp
@@ -1,5 +1,7 @@
 #include "cube.h"
 
+#define ZIPPATHDIV '/'
+
 enum
 {
     ZIP_LOCAL_FILE_SIGNATURE = 0x04034B50,
@@ -231,7 +233,7 @@ static void mountzip(ziparchive &arch, vector<zipfile> &files, const char *mount
         if(foundogz)
         {
             const char *ogzdir = foundogz;
-            while(--ogzdir >= f.name && *ogzdir != PATHDIV);
+            while(--ogzdir >= f.name && *ogzdir != ZIPPATHDIV);
             if(ogzdir < f.name || checkprefix(files, f.name, ogzdir + 1 - f.name))
             {
                 if(ogzdir >= f.name)
@@ -562,9 +564,14 @@ int listzipfiles(const char *dir, const char *ext, vector<char *> &files)
         {
             if(strncmp(f.name, dir, dirsize)) continue;
             const char *name = f.name + dirsize;
-            if(name[0] == PATHDIV) name++;
-            if(strchr(name, PATHDIV)) continue;
-            if(!ext) files.add(newstring(name));
+            if(name[0] == ZIPPATHDIV) name++;
+            const char *enddir = strchr(name, ZIPPATHDIV);
+            if(!ext)
+            {
+                if(enddir) files.add(newstring(name, enddir-name));
+                else files.add(newstring(name));
+            }
+            else if(enddir) continue;
             else
             {
                 size_t namelen = strlen(name);
