diff --git src/shared/zip.cpp src/shared/zip.cpp
index c35fa87..341e261 100644
--- src/shared/zip.cpp
+++ src/shared/zip.cpp
@@ -561,10 +561,11 @@ int listzipfiles(const char *dir, const char *ext, vector<char *> &files)
         enumerate(arch->files, zipfile, f,
         {
             if(strncmp(f.name, dir, dirsize)) continue;
-            const char *name = f.name + dirsize;
+            char *name = newstring(f.name + dirsize);
             if(name[0] == PATHDIV) name++;
-            if(strchr(name, PATHDIV)) continue;
-            if(!ext) files.add(newstring(name));
+            char *enddir = strchr(name, PATHDIV);
+            if(enddir) *enddir = '\0';
+            if(!ext) files.add(name);
             else
             {
                 size_t namelen = strlen(name);
