Index: src/Makefile
===================================================================
--- src/Makefile	(revision 6502)
+++ src/Makefile	(working copy)
@@ -57,6 +57,8 @@
   STRIP= $(TOOLCHAIN)strip
 endif
 endif
+CC= gcc-8 # gcc because of https://github.com/sauerbraten/p1xbraten/issues/26, 8 because that's the latest version available on AppVeyor
+CXX= gcc-8
 OSXMIN= 10.9
 override CC+= -arch x86_64 -mmacosx-version-min=$(OSXMIN)
 override CXX+= -arch x86_64 -mmacosx-version-min=$(OSXMIN)
@@ -63,7 +65,7 @@
 CLIENT_INCLUDES= $(INCLUDES) -Iinclude
 CLIENT_LIBS= -F../sauerbraten.app/Contents/Frameworks/ -framework SDL2 -framework SDL2_image
 CLIENT_LIBS+= -framework SDL2_mixer -framework CoreAudio -framework AudioToolbox
-CLIENT_LIBS+= -framework AudioUnit -framework OpenGL -framework Cocoa -lz -Lenet -lenet
+CLIENT_LIBS+= -framework AudioUnit -framework OpenGL -framework Cocoa -lz -Lenet -lenet -lc++
 else
 CLIENT_INCLUDES= $(INCLUDES) -I/usr/X11R6/include `sdl2-config --cflags`
 CLIENT_LIBS= -Lenet -lenet -L/usr/X11R6/lib -lX11 `sdl2-config --libs` -lSDL2_image -lSDL2_mixer -lz -lGL
@@ -142,6 +144,9 @@
 else
 SERVER_INCLUDES= -DSTANDALONE $(INCLUDES)
 SERVER_LIBS= -Lenet -lenet -lz
+ifneq (,$(findstring DARWIN,$(PLATFORM)))
+SERVER_LIBS+= -lc++
+endif
 MASTER_LIBS= $(SERVER_LIBS)
 endif
 SERVER_OBJS= \
@@ -221,6 +226,9 @@
 ifneq (,$(findstring DARWIN,$(PLATFORM)))
 install: client
 	cp sauer_client ../sauerbraten.app/Contents/MacOS/sauerbraten_universal
+ifneq (,$(STRIP))
+	$(STRIP) ../sauerbraten.app/Contents/MacOS/sauerbraten_universal
+endif
 else
 install: all
 	cp sauer_client	../bin_unix/$(PLATFORM_PREFIX)_client
