Index: src/Makefile
===================================================================
--- src/Makefile	(revision 6346)
+++ src/Makefile	(working copy)
@@ -232,6 +232,9 @@
 ifneq (,$(findstring DARWIN,$(PLATFORM)))
 install: client
 	cp sauer_client ../sauerbraten.app/Contents/MacOS/sauerbraten_universal
+ifneq (,$(STRIP))
+	$(STRIP) ../sauerbraten.app/Contents/MacOS/sauerbraten_universal
+endif
 else
 install: all
 	cp sauer_client	../bin_unix/$(PLATFORM_PREFIX)_client
