Index: src/p1xbraten/version.cpp
===================================================================
--- src/p1xbraten/version.cpp	(nonexistent)
+++ src/p1xbraten/version.cpp	(working copy)
@@ -15,6 +15,15 @@
     if(naturalsort(p1xbratenversion, version) == -1)
     {
         // we're newer, run migrations
+        if(naturalsort(p1xbratenversion, "9.0.0") == -1) {
+#ifndef STANDALONE
+            // unset CON_NONZEN (from archived/zenmode.patch) in all consoles
+            extern int confilter, fullconfilter, miniconfilter;
+            if((confilter&(1<<14)))         confilter -= 1<<14;
+            if((fullconfilter&(1<<14))) fullconfilter -= 1<<14;
+            if((miniconfilter&(1<<14))) miniconfilter -= 1<<14;
+#endif
+        }
     }
     setsvar("p1xbratenversion", version);
     lockversion();
