Index: src/engine/main.cpp
===================================================================
--- src/engine/main.cpp	(revision 6502)
+++ src/engine/main.cpp	(working copy)
@@ -1169,6 +1169,7 @@
 }

 VAR(numcpus, 1, 1, 16);
+VARP(minimizedframes, 0, 0, 1);

 int main(int argc, char **argv)
 {
@@ -1383,7 +1384,7 @@
         updateparticles();
         updatesounds();

-        if(minimized) continue;
+        if(minimized && !minimizedframes) continue;

         inbetweenframes = false;
         if(mainmenu) gl_drawmainmenu();
