Index: src/fpsgame/render.cpp
===================================================================
--- src/fpsgame/render.cpp	(revision 6291)
+++ src/fpsgame/render.cpp	(working copy)
@@ -226,6 +226,7 @@
 #endif

     VARP(statusicons, 0, 1, 1);
+    VARP(namesabovehead, 0, 1, 1);

     void renderstatusicons(fpsent *d, int team, float yoffset)
     {
@@ -238,12 +239,14 @@
             if(d->maxhealth>100) icons += (min(d->maxhealth, boost.max) - 100 + boost.info-1) / boost.info;
             if(d->armour>0 && d->armourtype>=A_GREEN && !m_noitems) icons++;
         }
-        if(icons) concatstring(d->info, " ");
-        particle_text(p, d->info, PART_TEXT, 1, team ? (team==1 ? 0x6496FF : 0xFF4B19) : 0x1EC850, 2.0f, 0, icons);
+        if (namesabovehead) {
+            if(icons) concatstring(d->info, " ");
+            particle_text(p, d->info, PART_TEXT, 1, team ? (team==1 ? 0x6496FF : 0xFF4B19) : 0x1EC850, 2.0f, 0, icons);
+        }
         if(icons)
         {
             float tw, th;
-            text_boundsf(d->info, tw, th);
+            text_boundsf(namesabovehead ? d->info : "", tw, th);
             float offset = (tw - icons*th)/2;
             if(d->armour>0 && d->armourtype>=A_GREEN && !m_noitems)
             {
