Index: src/engine/octaedit.cpp
===================================================================
--- src/engine/octaedit.cpp	(revision 6488)
+++ src/engine/octaedit.cpp	(working copy)
@@ -2517,6 +2517,7 @@
     intret(slot.variants->index);
 }

+ICOMMAND(settex, "i", (int *tex), { if(!vslots.inrange(*tex) || noedit()) return; filltexlist(); edittex(*tex); });
 COMMANDN(edittex, edittex_, "i");
 COMMAND(gettex, "");
 COMMAND(getcurtex, "");
